pkg/s3wrapper/util.go

```diff
@@ -70,24 +73,92 @@ func ExtractBootFilesFromISOAndUpload(ctx context.Context, log logrus.FieldLogge
 	return nil
 }
 
-func DownloadURLToTemporaryFile(url string) (string, error) {
-	tmpfile, err := ioutil.TempFile("", "isodownload")
-	if err != nil {
-		return "", errors.Wrap(err, "Error creating temporary file")
+func DownloadURLToFile(url string, isoObjectName string, checksum string) (string, error) {
+	filename := ""
+	isoDir := getEnv("CACHE_ISODIR", "/isos")
+	isoObjectPath := path.Join(isoDir, isoObjectName)
+	if stat, sterr := os.Stat(isoObjectPath); sterr == nil {
+		filename = path.Join(isoDir, stat.Name())
+		log.Infof("File already exists %s", filename)
+		valid, err := checkFileChecksum(isoObjectPath, checksum)
+		if err != nil {
+			return "", err
+		}
+		// If checksum don't match, remove the file from filesystem and download the new ISO.
+		if !valid {
+			if err = os.Remove(isoObjectPath); err != nil {
+				return "", err
+			}
+			filename, err = downloadISO(url, isoDir, isoObjectPath)
+			if err != nil {
+				return "", err
+			}
+		}
+	} else if os.IsNotExist(sterr) {
```

Comment:
I think this probably needs to be synchronized.
Check out https://github.com/openshift/assisted-service/blob/fabfeccc3eb4b536e52063b329fb649817f5dca2/pkg/s3wrapper/file_cache.go or https://github.com/openshift/assisted-service/blob/fabfeccc3eb4b536e52063b329fb649817f5dca2/internal/installercache/installercache.go for synchronized cache implementations.

Maybe it would be worth generalizing if we have three places we do this.